src/visualization.cpp

```diff
@@ -50,101 +49,130 @@ void Visualization::publishPlanPoint(const std_msgs::Header & header, const tf2:
-  mkSteps.id = __COUNTER__;
-  mkSteps.type = visualization_msgs::Marker::POINTS;
-  mkSteps.scale.x = 0.5;
-  mkSteps.scale.y = 0.5;
-  mkSteps.color.r = 1.0;
-  mkSteps.color.g = 0.5;
-  mkSteps.color.a = 1.0;
-  mkSteps.points = steps;
-
-  visualization_msgs::Marker mkPosesOnPath;
-  mkPosesOnPath.header.frame_id = frame_id;
-  mkPosesOnPath.header.stamp = ros::Time::now();
-  mkPosesOnPath.ns = "goal poses on path";
-  mkPosesOnPath.action = visualization_msgs::Marker::ADD;
-  mkPosesOnPath.pose.orientation.w = 1.0;
-  mkPosesOnPath.id = __COUNTER__;
-  mkPosesOnPath.type = visualization_msgs::Marker::POINTS;
-  mkPosesOnPath.scale.x = 0.5;
-  mkPosesOnPath.scale.y = 0.5;
-  mkPosesOnPath.color.r = 1.0;
-  mkPosesOnPath.color.g = 1.0;
-  mkPosesOnPath.color.a = 1.0;
-  mkPosesOnPath.points = path;
-
-  visualization_msgs::Marker mkCollisionFootprint;
-  mkCollisionFootprint.header.frame_id = frame_id;
-  mkCollisionFootprint.header.stamp = ros::Time::now();
-  mkCollisionFootprint.ns = "Collision footprint";
-  mkCollisionFootprint.action = visualization_msgs::Marker::ADD;
-  mkCollisionFootprint.pose.orientation.w = 1.0;
-  mkCollisionFootprint.id = __COUNTER__;
-  mkCollisionFootprint.type = visualization_msgs::Marker::LINE_LIST;
-  mkCollisionFootprint.scale.x = 0.1;
-  mkCollisionFootprint.color.b = 1.0;
-  mkCollisionFootprint.color.a = 0.3;
-  mkCollisionFootprint.points = footprint;
-
-  visualization_msgs::Marker mkCollisionHull;
-  mkCollisionHull.header.frame_id = frame_id;
-  mkCollisionHull.header.stamp = ros::Time::now();
-  mkCollisionHull.ns = "Collision polygon";
-  mkCollisionHull.action = visualization_msgs::Marker::ADD;
-  mkCollisionHull.pose.orientation.w = 1.0;
-  mkCollisionHull.id = __COUNTER__;
-  mkCollisionHull.type = visualization_msgs::Marker::LINE_STRIP;
-  mkCollisionHull.scale.x = 0.2;
-  mkCollisionHull.color.r = 1.0;
-  mkCollisionHull.color.a = 0.3;
-  mkCollisionHull.points = hull;
-
-  visualization_msgs::Marker mkCollisionIndicator;
-  mkCollisionIndicator.header.frame_id = frame_id;
-  mkCollisionIndicator.header.stamp = ros::Time::now();
-  mkCollisionIndicator.ns = "Collision object";
-  mkCollisionIndicator.pose.orientation.w = 1.0;
-  mkCollisionIndicator.id = __COUNTER__;
-  mkCollisionIndicator.type = visualization_msgs::Marker::CYLINDER;
-  mkCollisionIndicator.scale.x = 0.5;
-  mkCollisionIndicator.scale.y = 0.5;
-  mkCollisionIndicator.color.r = 1.0;
-  mkCollisionIndicator.color.a = 0.0;
-
-  visualization_msgs::MarkerArray mkCollision;
-  mkCollisionIndicator.scale.z = cost / 255.0;
-  mkCollisionIndicator.color.a = cost / 255.0;
-  mkCollisionIndicator.pose.position = point;
-  mkCollisionIndicator.pose.position.z = mkCollisionIndicator.scale.z * 0.5;
-  if (mkCollisionIndicator.scale.z > std::numeric_limits<float>::epsilon()) {
-    mkCollisionIndicator.action = visualization_msgs::Marker::ADD;
+  visualization_msgs::Marker marker;
+  marker.header = header;
```

Comment:
I do think  you missed out on an opportunity to reduce code duplication and magic values.

Personally, I really prefer the helper functions from my refactoring to create scales/colors/markers. And to create constants for colors/scales to reduce magic values.

Reply:
I've thought about this but I could not find a lot of duplicate code. Each markers color and scale is tuned by hand for visibility. Right now all values are local to the function so it's very readable.

Reply:
> I've thought about this but I could not find a lot of duplicate code. 

The following construct is repeated quite often:
```cpp
  visualization_msgs::Marker marker;
  marker.header = header;
  marker.ns = "extrapolated poses";
  marker.action = visualization_msgs::Marker::ADD;
  marker.pose.orientation.w = 1.0;
  marker.type = visualization_msgs::Marker::POINTS;
  marker.scale.x = 0.5;
  marker.scale.y = 0.5;
  marker.color.r = 1.0;
  marker.color.g = 0.5;
  marker.color.a = 1.0;
  marker.points = to_msg(steps);
  marker_pub_.publish(marker);
```

In my opinion, having a function return a marker to which you only pass the stuff that changes is better. Some advantages:
- the reader no longer has to figure out if the blocks are actually the same or not. its obvious
- if we decide we need to set/remove/change a data member we change the function and its impossible to forget a block
- its more obvious in which the blocks differ because we only have to check the arguments

> Each markers color and scale is tuned by hand for visibility. 

That's fine. It's not the actual values I have a problem with.

> Right now all values are local to the function so it's very readable.

I disagree. I find it hard to read. Because I see a lot of code which is very similar but not exactly. I suggest that you at least get rid of the magic values for colors and preferably also for scales.

If you have this
```cpp
  marker.color.g = 0.5;
  marker.color.a = 1.0;
```
The reader could wonder if green really was intended or that perhaps the r&b parts were forgotten. If you write the following (or something similar), it's immediately clear. Same is true for scales, IMHO.
```cpp
static const auto dark_green = ...{0,  0.5, 0, 1};
...
 marker.color = dark_green;
```

Reply:
I did the color name refactor as you asked. That helps a bit for readability.

For the rest of the marker, the blocks do look similar but they are different. That's why all those lines are there (doesn't violate DRY). I think this is the cleanest code we could get here. I think this
```c++
visualization_msgs::Marker marker;
marker.header = header;
marker.ns = "Collision object";
marker.type = visualization_msgs::Marker::CYLINDER;
marker.scale.x = 0.5;
marker.scale.y = 0.5;
marker.scale.z = cost / 255.0;
marker.color = red;
marker.color.a = cost / 255.0;
```
is easier to read then
```c++
auto marker_indicator = create_marker(
  header, "Collision object", visualization_msgs::Marker::CYLINDER,
  create_scale(0.5, 0.5, cost / 255.0),
  create_color(1.0, 0.0, 0.0, static_cast<float>(cost) / 255.0F), {});
```
because you don't immediately see what each argument does. Maybe this is personal preference though. With python you have keyword arguments which makes the caller side more clean, but we don't have that in C++ unfortunately. Is this good enough?